data-prepper-plugins/user-agent-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/useragent/UserAgentProcessor.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.processor.useragent;
+
+import org.opensearch.dataprepper.metrics.PluginMetrics;
+import org.opensearch.dataprepper.model.annotations.DataPrepperPlugin;
+import org.opensearch.dataprepper.model.annotations.DataPrepperPluginConstructor;
+import org.opensearch.dataprepper.model.event.Event;
+import org.opensearch.dataprepper.model.processor.AbstractProcessor;
+import org.opensearch.dataprepper.model.processor.Processor;
+import org.opensearch.dataprepper.model.record.Record;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import ua_parser.CachingParser;
+import ua_parser.Client;
+import ua_parser.Parser;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.opensearch.dataprepper.logging.DataPrepperMarkers.EVENT;
+
+@DataPrepperPlugin(name = "user_agent", pluginType = Processor.class, pluginConfigurationType = UserAgentProcessorConfig.class)
+public class UserAgentProcessor extends AbstractProcessor<Record<Event>, Record<Event>> {
+    private static final Logger LOG = LoggerFactory.getLogger(UserAgentProcessor.class);
+    private final UserAgentProcessorConfig config;
+    private final Parser userAgentParser;
+
+    @DataPrepperPluginConstructor
+    public UserAgentProcessor(final PluginMetrics pluginMetrics, final UserAgentProcessorConfig config) {
+        super(pluginMetrics);
+        this.config = config;
+        this.userAgentParser = new CachingParser();
```

Comment:
Are there any parameters we should configure here to ensure that the cache does not grow too large?

Reply:
Good catch! CachingParser() has a configurable cache size (default is 1000). I can expose it as one of the processor configuration options.

Reply:
With that size, I think this could come as a follow-on feature. You can create a GitHub issue and add this later. Or add now, I'm fine with either.

Reply:
Added `cache_size` as a processor option.